File: gof-design-patterns/srcs/P00_02_Inheritance/Dog.java
package P00_02_Inheritance;

// 基底クラス
class Animal {
    protected String name;
    
    public Animal(String name) {
        this.name = name;
    }
    
    public void makeSound() {
        System.out.println("動物の鳴き声");
    }
}

// 派生クラス
public class Dog extends Animal {
    public Dog(String name) {
        super(name);
    }
    
    @Override
    public void makeSound() {
        System.out.println(name + "がワンワンと鳴く");
    }

    public static void main(String[] args) {
        Dog dog = new Dog("ポチ");
        dog.makeSound();
    }
}

File: gof-design-patterns/typescripts/P00_02_Inheritance/Dog.ts
// 基底クラス
class Animal {
    protected name: string;  // protectedな変数
    
    constructor(name: string) {
        this.name = name;
    }
    
    makeSound(): void {
        console.log("動物の鳴き声");
    }
}

// 派生クラス
class Dog extends Animal {
    constructor(name: string) {
        super(name);  // 親クラスのコンストラクタを呼び出す
    }
    
    makeSound(): void {
        // メソッドのオーバーライド
        console.log(`${this.name}がワンワンと鳴く`);
    }
}

// 実行例
const dog = new Dog("ポチ");
dog.makeSound();
